server/src/main/java/org/elasticsearch/cluster/metadata/IndexNameExpressionResolver.java

```diff
@@ -674,7 +674,28 @@ boolean iterateIndexAliases(int indexAliasesSize, int resolvedExpressionsSize) {
                         }
                     }
                 }
-            } else {
+            } else if (indexAbstraction != null && indexAbstraction.getType() == IndexAbstraction.Type.DATA_STREAM) {
+                IndexAbstraction.DataStream dataStream = (IndexAbstraction.DataStream) indexAbstraction;
+                if (dataStream.getIndices() != null) {
+                    for (IndexMetadata indexMetadata : dataStream.getIndices()) {
+                        String concreteIndex = indexMetadata.getIndex().getName();
+                        if (norouting.contains(concreteIndex) == false) {
+                            norouting.add(concreteIndex);
+                            if (paramRouting != null) {
+                                Set<String> r = new HashSet<>(paramRouting);
+                                if (routings == null) {
+                                    routings = new HashMap<>();
+                                }
+                                routings.put(concreteIndex, r);
+                            } else {
+                                if (routings != null) {
+                                    routings.remove(concreteIndex);
+                                }
+                            }
+                        }
+                    }
+                }
+            }  else {
```

Comment:
I'll have to dig into this further, but shouldn't it _not_ resolve search routings if custom routing is not enabled on the data stream?

Reply:
done